ttnn/cpp/pybind11/pytensor.cpp

```diff
@@ -472,47 +476,64 @@ Tensor convert_python_tensors_to_tt_tensors(
 
 template <typename T>
 owned_buffer::Buffer<T> create_row_major_owned_buffer(
-    owned_buffer::Buffer<T> owned_buffer, const ttnn::TensorSpec& tensor_spec) {
-    if (tensor_spec.layout() == Layout::TILE) {
-        auto data = tensor_impl::convert_layout_tile_to_row_major(
-            tensor_spec.physical_shape(), tensor_spec.tile(), owned_buffer);
-        return owned_buffer::create(std::move(data));
+    owned_buffer::Buffer<T> owned_buffer, const ttnn::TensorSpec& tensor_spec, const bool legacy_output) {
+    if (tensor_spec.memory_config().is_sharded()) {
+        TT_FATAL(
+            tensor_spec.memory_config().shard_spec.has_value(),
+            "Sharded tensors must have a shard spec when converting from tt tensors!");
+    }
+
+    if (legacy_output) {
+        if (tensor_spec.layout() == Layout::TILE) {
+            auto data = tensor_impl::convert_layout_tile_to_row_major(
+                tensor_spec.physical_shape(), tensor_spec.tile(), owned_buffer);
+            return owned_buffer::create(std::move(data));
+        }
+        return owned_buffer;
     }
-    return owned_buffer;
+
+    auto physical_data = owned_buffer.get();
+
+    // Convert physical data based on tensor spec to row major logical data
+    auto logical_data = tensor_impl::convert_physical_data_to_logical_data(physical_data, tensor_spec);
+
+    return owned_buffer::create(std::move(logical_data));
 }
 
-std::variant<OwnedBuffer, BorrowedBuffer> get_host_buffer_from_tensor(const Tensor& tt_tensor) {
+std::variant<OwnedBuffer, BorrowedBuffer> get_host_buffer_from_tensor(
+    const Tensor& tt_tensor, const bool legacy_output) {
```

Comment:
what is `legacy_output`?

Reply:
I had to maintain a path to support old logic of returning physical data to torch tensor because some code depend on it. Off the top of my head, I think what broke was:
- ROW_MAJOR tensors with padding (since ROW_MAJOR has no alignment, I have no way to automatically strip it unless I query padded shape)
- Physical sharding (which is all sharding today) on padded shape. It doesn't work because (unlike the interleaved case), I can't derive an equivalent logical sharding to strip out data.

Reply:
Is there plan to get migrate the code off this path? What are the blockers / how difficult is it? Would also be great to add a TODO and describe what you described in the comment.

Reply:
I can add as a TODO, but no concrete plans yet

Reply:
I think we need to prioritize removing these legacy code paths... New code / new features are great, but maintaining old paths is just tech debt.

@ayerofieiev-tt

Reply:
I agree. I intend this to be the last PR I do for fully new features. I want to spend some time cleaning up as well. This last PR is high priority though since I can't tell op writers you can "create tensors with logical sharding" unless it works end-to-end with data.